File: app/core/socket.py
"""
Socket.io AsyncServer Setup

Architecture:
- One AsyncServer instance (singleton) shared across the app
- Room-based subscriptions: "auction:{id}" and "user:{id}"
- Clients join rooms to receive targeted events

Events (Client → Server):
  subscribe_auction   {"auction_id": int}  → join room "auction:{id}"
  unsubscribe_auction {"auction_id": int}  → leave room "auction:{id}"
  subscribe_user      {"user_id": int}     → join room "user:{id}"

Events (Server → Client):
  price_update        room="auction:{id}"  → {"auction_id", "current_price", "details", "timestamp"}
  turbo_triggered     room="auction:{id}"  → {"auction_id", "turbo_started_at", "remaining_minutes"}
  booking_confirmed   room="user:{id}"     → {"booking_code", "auction_id", "locked_price", "status"}
  auction_booked      room="auction:{id}"  → {"auction_id", "booking_code"}  (public: auction is taken)
"""

import socketio

# CORS origins accepted by the Socket.io server
_CORS_ORIGINS = ["http://localhost:3000", "http://localhost:8000", "*"]

# Singleton AsyncServer (async_mode="asgi" required for FastAPI/Starlette)
sio = socketio.AsyncServer(
    async_mode="asgi",
    cors_allowed_origins=_CORS_ORIGINS,
    logger=False,
    engineio_logger=False,
)


# ─────────────────────────────────────────────
# Connection lifecycle
# ─────────────────────────────────────────────

@sio.event
async def connect(sid: str, environ: dict, auth: dict = None):
    """Called when a client connects."""
    print(f"[Socket.io] Client connected: {sid}")


@sio.event
async def disconnect(sid: str):
    """Called when a client disconnects."""
    print(f"[Socket.io] Client disconnected: {sid}")


# ─────────────────────────────────────────────
# Room management events
# ─────────────────────────────────────────────

@sio.event
async def subscribe_auction(sid: str, data: dict):
    """
    Client asks to receive updates for a specific auction.
    Expected payload: {"auction_id": <int>}
    """
    auction_id = data.get("auction_id")
    if auction_id is None:
        await sio.emit("error", {"message": "auction_id required"}, to=sid)
        return
    room = f"auction:{auction_id}"
    await sio.enter_room(sid, room)
    await sio.emit("subscribed", {"room": room}, to=sid)
    print(f"[Socket.io] {sid} subscribed to {room}")


@sio.event
async def unsubscribe_auction(sid: str, data: dict):
    """Leave an auction room."""
    auction_id = data.get("auction_id")
    if auction_id is None:
        return
    room = f"auction:{auction_id}"
    await sio.leave_room(sid, room)
    await sio.emit("unsubscribed", {"room": room}, to=sid)


@sio.event
async def subscribe_user(sid: str, data: dict):
    """
    Client asks to receive personal notifications (booking confirmations, etc.).
    Expected payload: {"user_id": <int>}
    NOTE: In production, validate this with JWT – do not trust user-supplied user_id blindly.
    """
    user_id = data.get("user_id")
    if user_id is None:
        await sio.emit("error", {"message": "user_id required"}, to=sid)
        return
    room = f"user:{user_id}"
    await sio.enter_room(sid, room)
    await sio.emit("subscribed", {"room": room}, to=sid)
    print(f"[Socket.io] {sid} subscribed to {room}")

File: frontend/src/services/socket.js
import { io } from "socket.io-client";

// Using Vite environment variables (import.meta.env)
// If VITE_API_URL is not set, default to localhost:8000
const SOCKET_URL = import.meta.env.VITE_API_URL || "http://localhost:8000";

class SocketService {
  constructor() {
    this.socket = null;
    this.isConnected = false;
  }

  connect() {
    if (this.socket && this.socket.connected) return;

    this.socket = io(SOCKET_URL, {
      path: "/socket.io/",
      transports: ["websocket", "polling"],
      reconnectionAttempts: 5,
      reconnectionDelay: 2000,
    });

    this.socket.on("connect", () => {
      console.log("[SocketService] Connected:", this.socket.id);
      this.isConnected = true;
    });

    this.socket.on("disconnect", (reason) => {
      console.warn("[SocketService] Disconnected:", reason);
      this.isConnected = false;
    });

    this.socket.on("connect_error", (error) => {
      console.error("[SocketService] Connection Error:", error);
    });
  }

  disconnect() {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
      this.isConnected = false;
    }
  }

  // Room Subscription Methods
  subscribeAuction(auctionId) {
    if (!this.socket) return;
    console.log(`[SocketService] Subscribing to auction:${auctionId}`);
    this.socket.emit("subscribe_auction", { auction_id: auctionId });
  }

  unsubscribeAuction(auctionId) {
    if (!this.socket) return;
    console.log(`[SocketService] Unsubscribing from auction:${auctionId}`);
    this.socket.emit("unsubscribe_auction", { auction_id: auctionId });
  }

  subscribeUser(userId) {
    if (!this.socket) return;
    console.log(`[SocketService] Subscribing to user:${userId}`);
    this.socket.emit("subscribe_user", { user_id: userId });
  }

  // Event Listeners
  on(event, callback) {
    if (!this.socket) return;
    this.socket.on(event, callback);
  }

  off(event, callback) {
    if (!this.socket) return;
    this.socket.off(event, callback);
  }
}

export default new SocketService();
